### PR TITLE
[SYCL] Rework the way we handle SYCL -cc1 arguments.

### PR DESCRIPTION
**Pull request by AaronBallman**
-fsycl-is-device and -fsycl-is-host are mutually exclusive options that
now imply -fsycl mode. So passing -fsycl-is-device will now set both
the SYCL and SYCLIsDevice language options.

This introduces driver errors if the user attempts to pass both host
and device mode on the command line. Further, it disallows you from
specifying the -sycl-std= option unless in device or host mode.

There is some testing fallout from these changes because some tests
were passing -fsycl-is-device and not passing -fsycl which meant that
SYCL mode was never enabled for those tests.

### CODE REVIEW

**Comment by AaronBallman** [comment]
I wanted to get some early feedback on the approach I'm taking, but I believe this work should actually be landed upstream and then pulled in from there.

I believe there is one failing test case for clang/CodeGenSYCL/loop_unroll.cpp that I will likely need some help figuring out. The test was not actually testing in SYCL mode and so it was likely an invalid test to begin with.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (erichkeane) [inline clang/lib/Frontend/CompilerInvocation.cpp:465]: IMO, I'd prefer fsycl just NOT work in cc1 mode.

**Review by erichkeane — COMMENTED** [review]
1 comment that likely requires driver work, but I don't want fsycl to work in cc1 at all.  I think it should mean "do the 3 invocations thing" to the driver, and nothing to the cc1 invocation.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (AaronBallman) [reply clang/lib/Frontend/CompilerInvocation.cpp:465]: Do you know of any options in `Options.td` that do this? I tried to go down that route but I could find no way to get a driver-only option to work (mostly because it would never set `LangOpt.SYCL`). Related, does that mean you want `LangOpts.SYCL` to go away (because it's not a language option at that point, just a driver option)?

Comment (erichkeane) [reply clang/lib/Frontend/CompilerInvocation.cpp:465]: I looked through a couple, and see `OPT_driver_mode` which I think is only meaningful to the driver?  Same as `-c` or `-###`:

```
bash-4.2$ ./bin/clang --driver-mode=gcc c6.cpp -std=c++17 -c
bash-4.2$ ./bin/clang -cc1 --driver-mode=gcc c6.cpp -std=c++17 -c
error: unknown argument: '--driver-mode=gcc'
error: unknown argument: '-c'
```


I think `LangOpt.SYCL` is still useful as Host OR Device, but I'm not sure how often we care about 'any' vs a specific host/device.

Comment (erichkeane) [reply clang/lib/Frontend/CompilerInvocation.cpp:465]: As for the setting of `LangOpts.SYCL`, my thought is to make it just SYCLIsHost || SYCLIsDevice

Comment (AaronBallman) [reply clang/lib/Frontend/CompilerInvocation.cpp:465]: I think I actually prefer getting rid of `LangOpts.SYCL` entirely. It's used ~5 times in the frontend and only twice is it used to mean host or device mode. I think it's better to require people to think about host vs device instead of accidentally testing for SYCL in general.

Comment (erichkeane) [reply clang/lib/Frontend/CompilerInvocation.cpp:465]: SGTM!

Comment (AaronBallman) [reply clang/lib/Frontend/CompilerInvocation.cpp]: @erichkeane -- You asked for it, now you can have 219 changed files in a patch you'll be reviewing twice. ;-)

I still need some assistance with resolving the CodeGenSYCL/loop_unroll.cpp failure.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (erichkeane) [inline clang/lib/Frontend/CompilerInvocation.cpp:474]: Small preference for this to still say -fsycl' I think, since that is likely what the user is doing.

Comment (AaronBallman) [reply clang/lib/Frontend/CompilerInvocation.cpp:474]: The only downside I see to that is that we'll warn on: `clang -cc1 -sycl-std=2020 foo.cpp` by telling the user to pass `-fsycl` only for them to get a new diagnostic about `-fsycl` being an unknown option.

Perhaps a better approach would be for me to notice when the user does `clang -sycl-std=2020 foo.cpp` without specifying `-fsycl` and diagnose that from the driver side?

File: clang/include/clang/Driver/Options.td
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:4270]:
```suggestion
def fsycl : Flag<["-"], "fsycl">, Flags<[NoXarchOption, CoreOption]>, Group<sycl_Group>,
```
Needs to be a core option so it is available on Windows.  Also, should retain the negative variant if possible.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (erichkeane) [reply clang/lib/Frontend/CompilerInvocation.cpp:474]: That is probably optimal now that you mention it.

That said, users aren't typically expected to call our compiler with `-cc1` , right?  It is basically just an implementation detail.

File: clang/include/clang/Driver/Options.td
Comment (AaronBallman) [reply clang/include/clang/Driver/Options.td:4270]: Out of curiosity, why retain the negative variant?

File: clang/test/CodeGenSYCL/loop_unroll.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/loop_unroll.cpp:33]: I would prefer you pass -sycl-std=2020 on the run command instead.

File: clang/test/SemaSYCL/intel-fpga-mem-builtin.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/intel-fpga-mem-builtin.cpp:59]: Same here.

File: clang/test/SemaSYCL/intel-fpga-reg.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/intel-fpga-reg.cpp:60]: Same here.

File: clang/test/CodeGenSYCL/loop_unroll.cpp
Comment (AaronBallman) [reply clang/test/CodeGenSYCL/loop_unroll.cpp:33]: I'm fine doing that for all three tests, but FWIW, I considered doing that an explicitly decided against it because it seemed better to have test coverage for any SYCL language mode rather than making it seem like there's something 2020-specific about the test. I'm not certain if that changes your opinion about this or not. (let me know, as I said, I'm fine either way.)

File: clang/include/clang/Driver/Options.td
Comment (mdtoguchi) [reply clang/include/clang/Driver/Options.td:4270]: I see it more from a usability standpoint.  We have provided `-fno-sycl` to users so taking it away may have some impact.

Comment (AaronBallman) [reply clang/include/clang/Driver/Options.td:4270]: Okie dokie, I'll look into it.

File: clang/lib/Frontend/CompilerInvocation.cpp
Comment (AaronBallman) [reply clang/lib/Frontend/CompilerInvocation.cpp:474]: Neato, we test that `-sycl-std=` without `-fsycl` has a particular behavior rather than error out. See clang/test/Driver/sycl.c: `// RUN: %clang -### -sycl-std=2017 %s 2>&1 | FileCheck %s --check-prefix=NOT_ENABLED`

So should I change that test to expect an error, or should I assume that a user is relying on this behavior and remove the -cc1 diagnostic (so you can pass `-cc1 -sycl-std=2020` without specifying `-fsycl-is-host` or `-fsycl-is-device` and it's a noop)?

Comment (AaronBallman) [reply clang/lib/Frontend/CompilerInvocation.cpp:474]: (FWIW, I notice we also test this explicitly in CompilerInvocationTest.cpp for the unittests. So my inclination is to leave this be and not warn about it in -cc1.)

File: clang/test/CodeGenSYCL/loop_unroll.cpp
Comment (premanandrao) [reply clang/test/CodeGenSYCL/loop_unroll.cpp:33]: Doesn't really change my opinion, but now I won't mind if you didn't change it either.   I think these tests may eventually get changed to use the kernel_single_task function from the mock header, as opposed to this dummied up one.

**Comment by premanandrao** [comment]
> I believe there is one failing test case for clang/CodeGenSYCL/loop_unroll.cpp that I will likely need some help figuring out. The test was not actually testing in SYCL mode and so it was likely an invalid test to begin with.

My quick thought is that you should modify the test (which was invalid to begin with) to accept the new behavior with your change.

File: clang/test/CodeGenSYCL/loop_unroll.cpp
Comment (AaronBallman) [reply clang/test/CodeGenSYCL/loop_unroll.cpp:33]: Ohhh, I didn't realize this was an API as opposed to random test code. I see now why you asked for the changes. I'll switch the language standard mode.

**Review by premanandrao — APPROVED** [review]

**Review by mdtoguchi — DISMISSED** [review]

**Review by premanandrao — APPROVED** [review]

**Comment by AaronBallman** [comment]
Now that it looks like I've gotten everything to pass for us, I think the next steps are to figure out how to land this. I think these changes need to go upstream, so my intention is to spin up a community review for these changes. However, once this lands here through a pulldown, we'll have a *lot* of broken tests. My plan is to leave this PR languish as a Draft until pulldown breaks things, then resurrect it to fix the test failures. If anyone has suggestions on how to make it less painful for us, I'm all ears.

**Comment by AaronBallman** [comment]
Two things:

1) I have no idea what's going on with the Jenkins failures. The log files manage to contain both too much and too little information at the same time. Where do these tests live in the source tree? From the logs, I see `Cloning repository https://github.com/otcshare/llvm_ci.git` and yet... I can't seem to clone that repo.
2) When I started doing the upstreaming work, I found out that community has SYCL and SYCLIsDevice, but not SYCLIsHost. That makes upstreaming this... awkward, to say the least. I'm not certain how to proceed on the community side of things.

**Comment by AaronBallman** [comment]
The test case failures now make sense in light of offline discussions -- they're unrelated to the changes in this patch. The upstreaming questions remain.

**Comment by erichkeane** [comment]
IMO, this seems like a good excuse/point to upstream the SYCLIsHost existence, even if it does nothing, just so that the community command line all matches intel/llvm.

**Comment by premanandrao** [comment]
> Now that it looks like I've gotten everything to pass for us, I think the next steps are to figure out how to land this. 

Why can't you check it in here first and then upstream?   Will the subsequent pulldown cause too many failures?

**Comment by AaronBallman** [comment]
> > Now that it looks like I've gotten everything to pass for us, I think the next steps are to figure out how to land this.
> 
> Why can't you check it in here first and then upstream? Will the subsequent pulldown cause too many failures?

There's a possibility community will go "no no, do it like this instead" and we wind up with conflicts. However, I think checking it in here first is a good way forward given that upstream is less clear.

> IMO, this seems like a good excuse/point to upstream the SYCLIsHost existence, even if it does nothing, just so that the community command line all matches intel/llvm.

I think this is a good goal to at least explore (I sure hope it doesn't turn into yak shaving though).

**Review by premanandrao — DISMISSED** [review]
LGTM

**Comment by AaronBallman** [comment]
Folks are continuing to add new tests that add `-fsycl` to their `-cc1` command lines, so the longer this review remains unmerged, the more often I have to update it for these new broken tests.

File: clang/test/CodeGenSYCL/loop_unroll.cpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/loop_unroll.cpp:1]: @bader could you please take a look at the changes in this test to confirm behavior is as required. This test was incorrect since -fsycl wasn't passed earlier. The test started failing once all required options were set. If the test isn't behaving as expected, please file a bug report. I don't think we need to hold up this PR since fixing the test (if it needs to be) is outside the scope of this patch

File: clang/unittests/Frontend/CompilerInvocationTest.cpp
Comment (elizabethandrews) [inline clang/unittests/Frontend/CompilerInvocationTest.cpp:542]: -sycl-std*

Comment (elizabethandrews) [reply clang/unittests/Frontend/CompilerInvocationTest.cpp:542]: Just to confirm, Driver still passes -sycl-std whenever -fsycl is passed right?

**Review by elizabethandrews — APPROVED** [review]
Changes LGTM. I am not sure of changes in CompilerInvocationTest.cpp  because I am unfamiliar with this.

**Comment by AaronBallman** [comment]
@mdtoguchi, @AGindinson, or @intel/llvm-reviewers-runtime -- can you please take a look?

File: clang/unittests/Frontend/CompilerInvocationTest.cpp
Comment (AaronBallman) [reply clang/unittests/Frontend/CompilerInvocationTest.cpp:542]: Good catch on the typo, that's corrected now.

And yes, `-sycl-std=` gets passed when `-fsycl` gets passed. This is tested in test/Driver/sycl.c

**Review by elizabethandrews — DISMISSED** [review]

**Review by mdtoguchi — DISMISSED** [review]

File: clang/test/CodeGenSYCL/loop_unroll.cpp
Comment (bader) [reply clang/test/CodeGenSYCL/loop_unroll.cpp:1]: Changes looks okay to me, but it was added by @vmaksimo, so I would rather get her feedback instead.

I think setting `-sycl-std=2020` should be unnecessary as loop unroll metadata generation must be invariant to SYCL version.

Comment (vmaksimo) [reply clang/test/CodeGenSYCL/loop_unroll.cpp:1]: Looks okay, I agree with @bader

Comment (AaronBallman) [reply clang/test/CodeGenSYCL/loop_unroll.cpp]: I removed the `-sycl-std=2020` from the test, thanks for checking!

**Review by elizabethandrews — APPROVED** [review]

**Review by mdtoguchi — DISMISSED** [review]

**Review by premanandrao — DISMISSED** [review]

**Comment by bader** [comment]
@intel/llvm-reviewers-runtime, ping.

**Review by elizabethandrews — DISMISSED** [review]

**Comment by bader** [comment]
> @intel/llvm-reviewers-runtime, ping.

There is only one line of changes for you to review: sycl/test/check_device_code/id_queries_fit_int.cpp. :)

File: sycl/test/check_device_code/id_queries_fit_int.cpp
Comment (bader) [inline sycl/test/check_device_code/id_queries_fit_int.cpp:1]: It looks like this test is supposed to check LLVM IR emitted by the host compiler. Right?
1. Should we remove  `-fsycl` as well? Is there a better way to get the result of host-side compilation?
2. (not relevant to this PR) Should it be located in sycl/test/**check_device_code**/ directory, if it doesn't check device code at all?

Comment (AaronBallman) [reply sycl/test/check_device_code/id_queries_fit_int.cpp:1]: It was unclear what this test was trying to accomplish with that RUN line. `-fsycl` kicks off a device and a host build, so it wasn't obvious what was intended when running in host-and-device-at-the-same-time mode (which isn't supported).

If the test is only interested in host mode functionality, it should remove `-fsycl` and pass `-Xclang -fsycl-is-host`.

Comment (s-kanaev) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: > It looks like this test is supposed to check LLVM IR emitted by the host compiler. Right?

Exactly.

Comment (AaronBallman) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: > By the way, what's the purpose of this file change?

Passing `-fsycl` to the Clang driver spins off two frontend executions, one that passes `-fsycl-is-device` and one that passes `-fsycl-is-host`. The frontend can only ever be in one of those modes at a time (it's now an error to try to pass both flags). This test was passing `-fsycl -Xclang -fsycl-is-host`, so for the host execution everything was fine, but for the device execution would end up in both device and host mode.

Comment (AaronBallman) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: So then this is not a host-only test?

Comment (bader) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: @AaronBallman, driver doesn't add paths to DPC++ headers implicitly w/o `-fsycl` option, so we need either add `-I <path to DPC++ headers>` (there should be examples of this in other tests) or probably better option is to update the test and run it in "device-only" mode (I guess this mode implies `-fsycl` and driver add path).

Comment (AaronBallman) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: I'd prefer not to try rewriting the test (it seems unrelated to the changes in this PR -- the test was already questionable). I'll try adding the -I path to correct the issue, thanks for the suggestion!

**Review by mdtoguchi — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

File: sycl/test/check_device_code/id_queries_fit_int.cpp
Comment (bader) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: I agree. The test already had this flow, so it's better to address it separately.

Comment (AaronBallman) [reply sycl/test/check_device_code/id_queries_fit_int.cpp]: It seems that passing the path was insufficient to correct the test. Given that the test is already in a broken state, I'd prefer if someone with more domain knowledge handled correcting the test deficiencies in a separate PR (so I would back to the original changes to the test, which were originally passing).

Comment (s-kanaev) [reply sycl/test/check_device_code/id_queries_fit_int.cpp:1]: > To validate what @s-kanaev says is intended we need to re-write the code to make it device code instead.

Then, there must be an issue for us to not forget about it.

Comment (AaronBallman) [reply sycl/test/check_device_code/id_queries_fit_int.cpp:1]: @s-kanaev -- I filed https://github.com/intel/llvm/issues/3335 to track this.

Comment (s-kanaev) [reply sycl/test/check_device_code/id_queries_fit_int.cpp:1]: Great thanks!

**Pull request merged**
